File: java/src/test/java/money_problem/domain/BankTest.java
package money_problem.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static money_problem.domain.Currency.*;
import static money_problem.domain.DomainExtensions.dollars;
import static money_problem.domain.DomainExtensions.euros;
import static org.assertj.core.api.Assertions.assertThat;

class BankTest {
    private final Bank bank = Bank.withExchangeRate(EUR, USD, 1.2);

    @Test
    @DisplayName("10 EUR -> USD = 12 USD")
    void shouldConvertEuroToUsd() {
        assertThat(bank.convert(euros(10), USD)
                .get())
                .isEqualTo(dollars(12));
    }

    @Test
    @DisplayName("10 EUR -> EUR = 10 EUR")
    void shouldConvertInSameCurrency() {
        assertThat(bank.convert(euros(10), EUR)
                .get())
                .isEqualTo(euros(10));
    }

    @Test
    @DisplayName("Return a left in case of missing exchange rate")
    void shouldReturnALeftOnMissingExchangeRate() {
        assertThat(bank.convert(euros(10), KRW)
                .getLeft())
                .isEqualTo("EUR->KRW");
    }

    @Test
    @DisplayName("Conversion with different exchange rates EUR to USD")
    void shouldConvertWithDifferentExchangeRates() {
        assertThat(bank.convert(euros(10), USD)
                .get())
                .isEqualTo(dollars(12));

        assertThat(bank.addExchangeRate(EUR, USD, 1.3)
                .convert(euros(10), USD)
                .get())
                .isEqualTo(dollars(13));
    }
}

File: kotlin/src/test/kotlin/money_problem/domain/BankTest.kt
package money_problem.domain

import money_problem.domain.Currency.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class BankTest {
    private val bank = bankWithExchangeRate(EUR, USD, 1.2)

    @Test
    fun `10 EUR to USD = 12 USD`() {
        assertThat(
            bank.convert(10.0.euros(), USD)
                .rightUnsafe()
        ).isEqualTo(12.0.dollars())
    }

    @Test
    fun `10 EUR to EUR = 10 EUR`() {
        assertThat(
            bank.convert(10.0.euros(), EUR)
                .rightUnsafe()
        ).isEqualTo(10.0.euros())
    }

    @Test
    fun `Return a Left in case of missing exchange rates`() {
        assertThat(
            bank.convert(10.0.euros(), KRW)
                .leftUnsafe()
        ).isEqualTo("EUR->KRW")
    }

    @Test
    fun `Conversion with different exchange rates EUR to USD`() {
        assertThat(
            bank.convert(10.0.euros(), USD)
                .rightUnsafe()
        ).isEqualTo(12.0.dollars())

        assertThat(
            bank.addExchangeRate(EUR, USD, 1.3)
                .convert(10.0.euros(), USD)
                .rightUnsafe()
        ).isEqualTo(13.0.dollars())

    }
}